packages/common/infra/src/di/__tests__/di.spec.ts

```diff
@@ -0,0 +1,357 @@
+        const command = this.commands.find(c => c.shortcut === shortcut);
+        if (command) {
+          command.callback();
+        }
+      }
+    }
+
+    serviceCollection.add(PageSystem);
+    serviceCollection.add(CommandSystem, [[Command]]);
+    serviceCollection.addImpl(Command('switch'), p => ({
+      shortcut: 'option+s',
+      callback: () => p.get(PageSystem).switchToEdgeless(),
+    }));
+    serviceCollection.addImpl(Command('rename'), p => ({
+      shortcut: 'f2',
+      callback: () => p.get(PageSystem).rename(),
+    }));
+
+    const provider = serviceCollection.provider();
+    const commandSystem = provider.get(CommandSystem);
+
+    expect(
+      pageSystemInitialized,
+      "PageSystem won't be initialized until command executed"
+    ).toEqual(false);
+
+    commandSystem.execute('option+s');
+    expect(pageSystemInitialized).toEqual(true);
+    expect(provider.get(PageSystem).mode).toEqual('edgeless');
+
+    expect(provider.get(PageSystem).name).toEqual('helloworld');
+    expect(commandSystem.commands.length).toEqual(2);
+    commandSystem.execute('f2');
+    expect(provider.get(PageSystem).name).toEqual('foobar');
+  });
+
+  test('duplicate, override', () => {
+    const serviceCollection = new ServiceCollection();
+
+    const something = createIdentifier<any>('USB');
+
+    class A {
+      a = 'i am A';
+    }
+
+    class B {
+      b = 'i am B';
+    }
+
+    serviceCollection.addImpl(something, A).override(something, B);
+
+    const provider = serviceCollection.provider();
+    expect(provider.get(something)).toEqual({ b: 'i am B' });
+  });
+
+  test('scope', () => {
+    const services = new ServiceCollection();
+
+    const workspaceScope = createScope('workspace');
+    const pageScope = createScope('page', workspaceScope);
+    const editorScope = createScope('editor', pageScope);
+
+    class System {
+      appName = 'affine';
+    }
+
+    services.add(System);
+
+    class Workspace {
+      name = 'workspace';
+      constructor(public system: System) {}
+    }
+
+    services.scope(workspaceScope).add(Workspace, [System]);
+    class Page {
+      name = 'page';
+      constructor(
+        public system: System,
+        public workspace: Workspace
+      ) {}
+    }
+
+    services.scope(pageScope).add(Page, [System, Workspace]);
+
+    class Editor {
+      name = 'editor';
+      constructor(public page: Page) {}
+    }
+
+    services.scope(editorScope).add(Editor, [Page]);
+
+    const root = services.provider();
+    expect(root.get(System).appName).toEqual('affine');
+    expect(() => root.get(Workspace)).toThrowError(ServiceNotFoundError);
+
+    const workspace = services.provider(workspaceScope, root);
+    expect(workspace.get(Workspace).name).toEqual('workspace');
+    expect(workspace.get(System).appName).toEqual('affine');
+    expect(() => root.get(Page)).toThrowError(ServiceNotFoundError);
+
+    const page = services.provider(pageScope, workspace);
```

Comment:
since we explictly pass `workspace` as `page` scope's parent when creating it, is it possible to ignore the parent scope parameter when getting the provider?

Reply:
No, the workspace will have multiple instances, and it must be clear which instance the page is derived from.